src/util.cc

```diff
@@ -423,21 +427,79 @@ const char* SpellcheckString(const char* text, ...) {
 }
 
 #ifdef _WIN32
+wstring Utf8ToWide(const string& u8_string) {
+  if (u8_string.empty())
+    return wstring();
+
+  size_t size_needed = MultiByteToWideChar(CP_UTF8, 0, u8_string.data(),
+                                           (int)u8_string.size(), NULL, 0);
+  if (size_needed == 0)
+    Fatal("Failed conversion to wide string");
+
+  vector<wchar_t> w_buffer(size_needed);
+  int w_size_needed =
+      MultiByteToWideChar(CP_UTF8, 0, u8_string.data(), (int)u8_string.size(),
+                          &w_buffer[0], w_buffer.size());
+  if (w_size_needed == 0)
+    Fatal("Failed conversion to wide string");
+
+  return std::wstring(&w_buffer[0], w_size_needed);
+}
+
+string WideToUtf8(const wstring& w_string) {
+  if (w_string.empty())
+    return std::string();
+
+  int size_needed = WideCharToMultiByte(
+      CP_UTF8, 0, &w_string[0], (int)w_string.size(), NULL, 0, NULL, NULL);
+  if (size_needed == 0)
+    Fatal("Failed conversion to UTF-8 string");
+
+  std::string u8_string(size_needed, 0);
+  WideCharToMultiByte(CP_UTF8, 0, &w_string[0], (int)w_string.size(),
+                      &u8_string[0], size_needed, NULL, NULL);
+  if (size_needed == 0)
+    Fatal("Failed conversion to UTF-8 string");
+
+  return u8_string;
+}
+
+char** convertCommandLine(int argc, wchar_t** wargv) {
+  wstring w_argString;
+  string argString;
+  char** argv = new char*[argc + 1];
+
+  for (int i = 0; i < argc; i++) {
+    // Convert the string
+    argString = WideToUtf8(wargv[i]);
+    // Prepare space in the vector
+    argv[i] = new char[argString.length() + 1];
+
+    // Copy the final results.
+    strcpy(argv[i], argString.c_str());
+  }
+
+  // Just to follow the c++ standard, we need to add this!
+  // Also, ninja will not work properly if we do not do so.
+  argv[argc] = NULL;
+  return argv;
+}
+
 string GetLastErrorString() {
   DWORD err = GetLastError();
 
-  char* msg_buf;
-  FormatMessageA(
+  wchar_t* msg_buf;
```

Comment:
Only do this on a windows platform.

Really, Linux / Mac never ever use wchar_t.

The windows specific quirks need to be better encapsulated so that they don't infect non-windows parts of the code.

Maybe make a "PlatformSpecificString" type that holds a single string type by value, and on Windows that type is wchar_t, and on every other platform, it's char. The conversion to the appropriate platform type can then happen just once, and the various areas of the Ninja codebase that care about the string width can be templated based on the PlatformSpecificString type's value_type typedef.

Reply:
This is a windows only function. Am I missing something?

Reply:
My suggestion is invalid. I did not see the #ifdef _WIN32.